server/src/main/java/io/deephaven/server/object/ObjectServiceGrpcImpl.java

```diff
@@ -50,6 +44,92 @@ public ObjectServiceGrpcImpl(SessionService sessionService, TicketRouter ticketR
+                // Should only appear in the first request
+                if (this.object != null) {
+                    throw Exceptions.statusRuntimeException(Code.FAILED_PRECONDITION,
+                            "Already sent a connect request, cannot send another");
+                }
+
+                TypedTicket typedTicket = request.getConnect().getTypedTicket();
+
+                final String type = typedTicket.getType();
+                if (type.isEmpty()) {
+                    throw Exceptions.statusRuntimeException(Code.INVALID_ARGUMENT, "No type supplied");
+                }
+                if (typedTicket.getTicket().getTicket().isEmpty()) {
+                    throw Exceptions.statusRuntimeException(Code.INVALID_ARGUMENT, "No ticket supplied");
+                }
+
+                final SessionState.ExportObject<Object> object = ticketRouter.resolve(
+                        session, typedTicket.getTicket(), "sourceId");
+
+                this.object = object;
+                manage(this.object);
+                session.nonExport()
+                        .require(object)
+                        .onError(responseObserver)
+                        .submit(() -> {
+                            final Object o = object.get();
+                            final ObjectType objectType = getObjectTypeInstance(type, o);
+
+                            messageStream =
+                                    objectType.clientConnection(o, new PluginMessageSender(responseObserver, session));
+                        });
+            } else if (request.hasData()) {
+                // All other requests
+                Data data = request.getData();
+                List<SessionState.ExportObject<Object>> referenceObjects = data.getTypedExportIdsList().stream()
+                        .map(typedTicket -> ticketRouter.resolve(session, typedTicket.getTicket(), "messageObjectId"))
+                        .collect(Collectors.toList());
+                List<SessionState.ExportObject<Object>> requireObjects = new ArrayList<>(referenceObjects);
+                requireObjects.add(object);
+                session.nonExport()
+                        .require(requireObjects)
+                        .onError(responseObserver)
+                        .submit(() -> {
+                            Object[] objs = referenceObjects.stream().map(ExportObject::get).toArray();
+                            // TODO: Should we try/catch this to recover/not close from plugin errors? Or make the
+                            // client
+                            // recover? Would be useful if the plugin is complex
+                            messageStream.onData(data.getPayload().asReadOnlyByteBuffer(), objs);
+                        });
+            }
+        }
+
+        @Override
+        public void onError(final Throwable t) {
+            // Do we need to clean up export collector here and onCompleted too?
+            release();
```

Comment:
We discussed that this is incomplete as-is. Needs to clear the queue, close the messageStream if it's non-null, and *then* release.